servers/physics_2d/physics_2d_server_sw.cpp

```diff
@@ -167,23 +167,14 @@ void Physics2DServerSW::_shape_col_cbk(const Vector2 &p_point_A, const Vector2 &
 	if (cbk->max == 0)
 		return;
 
-	if (cbk->valid_dir != Vector2()) {
+	Vector2 rel_dir = (p_point_A - p_point_B).normalized();
+	if (cbk->valid_depth < 10e20 && cbk->valid_dir != Vector2() && rel_dir != Vector2()) {
 		if (p_point_A.distance_squared_to(p_point_B) > cbk->valid_depth * cbk->valid_depth) {
 			cbk->invalid_by_dir++;
 			return;
 		}
```

Comment:
?
```c++
	Vector2 rel_dir = p_point_A - p_point_B;
	if (rel_dir.length_squared() > 0) {
		if (cbk->valid_depth < 10e20 && cbk->valid_dir != Vector2()) {
			if (p_point_A.distance_squared_to(p_point_B) > cbk->valid_depth * cbk->valid_depth) {
				cbk->invalid_by_dir++;
				return;
			}
			if (cbk->valid_dir.dot(rel_dir.normalized()) <= 0) {
				cbk->invalid_by_dir++;
				return;
			}
		}
	}
```

Reply:
I agree that optimization is possible. My goal in this PR is to keep the changes to a minimum, which makes it easier to review, and, therefore, more likely to be merged.

Reply:
It's the same, personally I just prefer to avoid normalizing null vectors as much as possible as it often leads to errors that are difficult to spot.

Reply:
Updated.